### ISSUE TITLE
ClassGraph loaded class missing annotations (regression)

### ISSUE THREAD

**Issue by TheSpiritXIII** [classgraph/classgraph#485]
I have an application that is functionally equivalent to the given example below. I have:
1. A Spring Boot application that uses ClassGraph to find uses of an annotation.
2. A Java application that uses ClassGraph to scan the Spring Boot uber-JAR.

When Java application scans my Spring Boot application, it finds the annotations via `AnnotationInfo` but `loadClass` gives me a `Class<?>` which has no annotations.

If I use ClassGraph 4.8.66, it works. If I run my Spring Boot application independently, it also works.

I have tried with `verbose` on each level and both find the annotation. I tried to compare runs on `4.8.66` and `4.8.67` but the logs were equivalent.

The example involves a bit of setup. I have setup a minimal Gradle multi-build project:
`build.gradle`
```gradle
plugins {
	id 'org.springframework.boot' version '2.3.3.RELEASE' apply false
}

subprojects {
	apply plugin: 'java'
	apply plugin: 'maven-publish'

	repositories {
		mavenLocal()
		jcenter()
	}

	publishing {
		repositories {
			publications {
				mavenJava(MavenPublication) {
					from components.java
				}
			}
		}
	}

	sourceCompatibility = '1.11'
	targetCompatibility = '1.11'

	// Works:
	def classGraphVersion = '4.8.66'
	// Does not work (comment out the following line to test working case):
	classGraphVersion = '4.8.92'

	dependencies {
		implementation group: 'io.github.classgraph', name: 'classgraph', version: classGraphVersion
	}
}
```

`gradle.properties`
```gradle
version = 1.0.0
group = com.thespiritxiii.example
```

`settings.gradle`
```gradle
rootProject.name = 'example'
include 'spring'
include 'scanner'
```

`spring/build.gradle`
```gradle
apply plugin: 'org.springframework.boot'

dependencies {
	implementation platform(org.springframework.boot.gradle.plugin.SpringBootPlugin.BOM_COORDINATES)
	implementation group: 'org.springframework.boot', name: 'spring-boot-starter-web'
}
```

`spring/src/main/java/example/Application.java`
```java
package example;

import java.lang.annotation.Annotation;
import java.util.function.Function;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

import io.github.classgraph.ClassGraph;
import io.github.classgraph.ClassInfo;
import io.github.classgraph.ScanResult;

@SpringBootApplication
public class Application {
	public static void main(String[] args) {
		System.out.println("Annotation count: " + find());
		SpringApplication.run(Application.class, args);
	}

	private static <A extends Annotation, T> int findGeneric(
		Class<A> annotationClass,
		Function<A, T> getValue
	) {
		int count = 0;
		final ClassGraph classGraph = new ClassGraph()
			.enableAllInfo()
			.enableInterClassDependencies()
			.enableExternalClasses();

		try (ScanResult scanResult = classGraph.enableAnnotationInfo().scan()) {
			for (ClassInfo classInfo : scanResult.getClassesWithAnnotation(annotationClass.getName())) {
				final Class<?> classType = classInfo.loadClass();
				final A mixin = classType.getAnnotation(annotationClass);
				// This should never be null because I called getClassesWithAnnotation with the same class.
				if (mixin == null) {
					throw new RuntimeException("Unable to find annotation on class: " + classType);
				}
				final T value = getValue.apply(mixin);
				if (value != null) {
					count += 1;
				}
			}
		}

		return count;
	}

	public static final int find() {
		return findGeneric(SomeAnnotation.class, SomeAnnotation::value);
	}
}
```

`spring/src/main/java/example/SomeAnnotation.java`
```java
package example;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface SomeAnnotation {
	String value();
}
```

`spring/src/main/java/example/SomeAnnotatedClass.java`
```java
package example;

@SomeAnnotation("String")
public class SomeAnnotatedClass {
}
```

`scanner/build.gradle`
```gradle
apply plugin: 'application'

application {
	mainClassName = 'example.App'
}

dependencies {
	// Spring Boot builds as an Uber jar so we don't need transitive dependencies.
	implementation(group: 'com.thespiritxiii.example', name: 'spring', version: project.version) {
		transitive = false
	}
}
```

`scanner/src/main/java/example/App.java`
```java
package example;

import java.lang.reflect.Method;
import java.util.concurrent.ForkJoinPool;

import io.github.classgraph.ClassGraph;
import io.github.classgraph.ScanResult;

public class App {
	public static void main(String[] args) {
		final ClassGraph classGraph = new ClassGraph()
			.enableAllInfo()
			.enableExternalClasses()
			.enableInterClassDependencies();

		// Run with 1 core (deterministically).
		try (ScanResult scanResult = classGraph.scanAsync(ForkJoinPool.commonPool(), 1).get()) {
			final Class<?> classType = scanResult.getClassInfo("example.Application").loadClass();
			final Method method = classType.getMethod("find");
			final Object value = method.invoke(null);
			System.out.println("Value is: " + value);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
}
```

To test:
```bash
./gradlew spring:bootRun
```
Outputs 1.

```bash
./gradlew spring:build spring:publishToMavenLocal scanner:run
```
Throws an exception if ClassGraph is 4.8.67 or above. If 4.8.66, outputs 1.

After testing, delete `~/.m2/repository/com/thespiritxiii/example/spring` from the maven local publish task.

**Comment by lukehutch** [comment]
Hi,

I tried recreating this project, and I couldn't get Eclipse to build it. From the commandline, the command `./gradlew spring:bootRun` appears to get stuck at `75% EXECUTING`.

Rather than paste the contents of each file here, can you please upload the complete project? You can just drag the zipped project onto the GitHub bug comment pane to attach it.

Although even before that, to save time (since you're way more familiar with gradle than I am), can you please try using `git bisect` with ClassGraph to find the commit that breaks this code for you?

Thank you!

**Comment by TheSpiritXIII** [comment]
Hey,

It's not stuck. Spring Boot is a web server. It doesn't ever finish hence why it says it's still executing. The relevant log line should be all the way at the top. Although, you can just skip the first command and run the second which is the one that really demonstrates the issue.

Apologies, but my computer's corporate policy prevents me from uploading zip files.

As for the issue, it is starts happening somewhere in between 4.8.66 and 4.8.67. I did check on the commits. I'm not sure why this would cause issues and it's certainly possible that the code base (and issue) changed since then:
https://github.com/classgraph/classgraph/compare/690da39f6b5e8d3f9e6aa2d52faf2922c6960d15...cadb501c847d6876727a8aed68aeecaa6487f107

**Comment by lukehutch** [comment]
I just tried re-creating the project, and got this in the last step. Not sure why I didn't get this last time:

```
$ ./gradlew spring:build spring:publishToMavenLocal scanner:run
> Task :spring:publishMavenJavaPublicationToMavenLocal FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':spring:publishMavenJavaPublicationToMavenLocal'.
> Failed to publish publication 'mavenJava' to repository 'mavenLocal'
   > Artifact spring-1.0.0.jar wasn't produced by this build.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
5 actionable tasks: 5 executed
```

Hmm, unfortunately some of the classloader changes I had to make between 4.8.66 and 4.8.67 were to fix classloading for other users. This may mean that Spring-Boot needs some special-case handling or something (highly likely, actually).

**Comment by lukehutch** [comment]
Also, is this broken for you on both JDK 8 and JDK 11? (or other pre- and post-JPMS JDK)

**Comment by TheSpiritXIII** [comment]
I looked into your problem and it looks like it's due to Gradle 6. At my company, we're still using Gradle 5. You can probably downgrade with:
```bash
./gradlew wrapper --gradle-version=5.6.4
```

Luckily, I still had JDK 8 installed on my machine. I tested it and I get the same results -- working in 4.8.66 but broken in 4.8.67.

**Comment by TheSpiritXIII** [comment]
Got it working on Gradle 6. Update `spring/build.gradle` with:
```gradle
apply plugin: 'org.springframework.boot'

configurations {
	[apiElements, runtimeElements].each {
		it.outgoing.artifacts.removeIf { it.buildDependencies.getDependencies(null).contains(jar) }
		it.outgoing.artifact(bootJar)
	}
}

dependencies {
	implementation platform(org.springframework.boot.gradle.plugin.SpringBootPlugin.BOM_COORDINATES)
	implementation group: 'org.springframework.boot', name: 'spring-boot-starter-web'
}
```

**Comment by lukehutch** [comment]
Thanks, I can get this to fail on the commandline now. Weirdly though, if I run it on Spring Tools in Eclipse, it succeeds with "Annotation count: 1". Any idea what is different about the Spring Tools environment? I was hoping to debug this in Eclipse, so this is a problem...

I searched for and tried all sorts of ways of specifying JVM options in Gradle, but can't figure out how to get this to launch from the commandline in debug mode with `suspend=y`, so that I can connect a remote debugging session in Eclipse, while pausing on entry to the application. Any suggestions?

**Comment by TheSpiritXIII** [comment]
It's a multi-project build. There's the `spring` project and the `runner` project. Spring Tools may be running the former. The problem is in the latter which is a normal Java application, i.e. not Spring.

According to the Gradle docs, you can add this to the application block:
`scanner/build.gradle`
```gradle
application {
	mainClassName = 'example.App'
	applicationDefaultJvmArgs = [ '-Xdebug', '-Xrunjdwp:transport=dt_socket,server=y,suspend=y,address=5005' ]
}
```

Supposedly, there's also an easier `--debug-jvm` argument you can try.
```gradle
./gradlew scanner:run --debug-jvm
```

**Comment by lukehutch** [comment]
I had tried both of those already. I can't get the suspend part to work, so the server runs and then dies before I can stop it. I'll try adding a `Thread.sleep` to `main`, but any other ideas?

**Comment by lukehutch** [comment]
Oh! Eclipse added a new "Stop in main" option to the remote debug launcher. Pretty sure it wasn't there last time I tried to do something like this. Debugging this now.

**Comment by lukehutch** [comment]
What I have found so far: in the broken version, the classes have `AppClassLoader` as their classloader (in JDK 11+). This fails to load the class, because the package root is not at the root of the jarfile, so `ClassGraphClassLoader` manually loads the class using `defineClass(className, resourceByteBuffer, null)`, where `resourceByteBuffer` contains the bytes of the classfile. Strangely the class defined in this way does not have any annotations -- presumably `AppClassLoader` tries loading the annotation classes itself rather than delegating to `ClassGraphClassLoader`, and fails. I don't know why `ClassGraphClassLoader` is not called to try to load the annotation classes...

**Comment by lukehutch** [comment]
I found this:

>To create new custom class loaders, the process is similar in the Java SE 7 release as in previous releases. Create a subclass of ClassLoader, then override the findClass() method and possibly loadClass(). Overriding loadClass() makes your life more difficult, but it is the only way to use a different delegation model.

So far `ClassGraphClassLoader` only implements `findClass`. Looks like I'll need to also override `loadClass` to support loading classes from Spring-Boot jarfiles when not running a Spring-Boot application.

**Comment by lukehutch** [comment]
OK, I figured out what's going on here...

Both `App.java` and `Application.java` produce an instance of `ScanResult`, and every `ScanResult` creates its own `ClassGraphClassLoader`, which is important for lifecycle management, since retaining a reference to a classloader instance can be a source of resource leaks.

`App.java` loads the `example.Application` class with the first `ClassGraphClassLoader`. This is used to call `Application.find`. Then `Application.java` uses its own classloader to resolve the direct annotation class reference `SomeAnnotation.class`. This is loaded using the first `ClassGraphClassLoader`.

However, `Application.java` then runs its own scan, loading `example.SomeAnnotatedClass` using the **second** `ClassGraphClassLoader` (from the second `ScanResult`). So when you call `mixin = classType.getAnnotation(annotationClass)`, the annotation cannot be found, because `annotationClass` has a different classloader than `classType`.

So this is not actually a bug in ClassGraph -- it is doing what it is supposed to do here. Presumably the fact that this worked before 4.8.66 was that all `ScanResult` instances were sharing a single `ClassGraphClassLoader` instance -- I know at one point I fixed a bug report of a resource leak of `ClassLoader` instances, and I think this may have been when it was fixed (I didn't check for sure).

Your issue here can be fixed by loading the `SomeAnnotation` class by name, rather than by reference:

```
                final Class<A> annotationClassType = (Class<A>) scanResult.getClassInfo("example.SomeAnnotation")
                        .loadClass();
                final A mixin = classType.getAnnotation(annotationClassType);
```

However this throws a new exception:

```
Starting scan
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at example.App.main(App.java:23)
Caused by: java.lang.ClassCastException: class com.sun.proxy.$Proxy3 cannot be cast to class example.SomeAnnotation (com.sun.proxy.$Proxy3 is in unnamed module of loader io.github.classgraph.ClassGraphClassLoader @5fbb5bef; example.SomeAnnotation is in unnamed module of loader io.github.classgraph.ClassGraphClassLoader @2133c8f8)
        at example.Application.findGeneric(Application.java:38)
        at example.Application.find(Application.java:49)
        ... 5 more
```

This is triggered by this line:

```
                final T value = getValue.apply(mixin);
```

The problem here is that `getValue` was again obtained from a direct class reference: `SomeAnnotation::value`, so the `InvocationHandler` used to mock up the annotation class (i.e. `$Proxy3`) is not compatible with the second `SomeAnnotation` class, therefore the methods are not compatible either. (An `InvocationHandler` has to be used by ClassGraph to load annotation classes, since concrete annotations cannot be instantiated directly.)

This can be fixed by replacing this with another reflective access by name:

```
@SpringBootApplication
public class Application {
    public static void main(String[] args) {
        System.out.println("Annotation count: " + find());
        SpringApplication.run(Application.class, args);
    }

    private static <A extends Annotation, T> int findGeneric(String annotationClassName,
            String annotationValueMethodName) {
        int count = 0;
        final ClassGraph classGraph = new ClassGraph().enableAllInfo().enableInterClassDependencies()
                .enableExternalClasses();

        try (ScanResult scanResult = classGraph.enableAnnotationInfo().scan()) {
            for (ClassInfo classInfo : scanResult.getClassesWithAnnotation(annotationClassName)) {
                final Class<?> classType = classInfo.loadClass();
                @SuppressWarnings("unchecked")
                final Class<A> annotationClassType = (Class<A>) scanResult.getClassInfo(annotationClassName)
                        .loadClass();
                final A mixin = classType.getAnnotation(annotationClassType);
                if (mixin == null) {
                    throw new RuntimeException("Unable to find annotation on class: " + classType);
                }
                try {
                    @SuppressWarnings("unchecked")
                    T value = (T) annotationClassType.getMethod(annotationValueMethodName).invoke(mixin);
                    if (value != null) {
                        count += 1;
                    }
                } catch (NoSuchMethodException | SecurityException | IllegalAccessException
                        | IllegalArgumentException | InvocationTargetException e) {
                    throw new RuntimeException(e);
                }
            }
        }

        return count;
    }

    public static final int find() {
        return findGeneric("example.SomeAnnotation", "value");
    }
}
```

**Comment by lukehutch** [comment]
That said, this exposed another issue here. I thought that this could be fixed by simply using the following in `findGeneric`:

```
        try (ScanResult scanResult = classGraph.enableAnnotationInfo()
                .overrideClassLoaders(Application.class.getClassLoader())   // Added line
                .scan()) {
            //...
        }
```

This would mean that both `ScanResults` ended up sharing (or in the second case, delegating to) a single `ClassGraphClassLoader` instance -- but since `ClassGraphClassLoader` does not extend `URLClassLoader` or any other supported classloader, ClassGraph can't find its own classpath URLs from its own classloader. I'll fix that so that the above pattern is supported.

**Comment by lukehutch** [comment]
So `.overrideClassLoaders(Application.class.getClassLoader())` is now supported as an easier way to solve the problem you ran into.

Note however that sharing a `ClassGraphClassLoader` between nested scans is not advisable, because scan criteria may differ between the outer scan and the inner scan.

If you do override the classloader in this way, the outer nested scan's classloader will be deferred to first, followed by attempting to load classes from the inner nested scan's classloader. This should work to an arbitrary nesting depth.

You'll need to modify your code using one of the methods I showed above, but this is fixed in 4.8.93.

Thanks for the bug report, and for being responsive with answers!

**Comment by TheSpiritXIII** [comment]
Thank you for your through investigation and prompt replies! That solution should work for me. :)
